### PR TITLE
Loop and drop second interaction

### PR DESCRIPTION
**Pull request by bioshazard**
## Why are these changes needed?

I am using `autogen` with **MistralOrca 7B 8k** and ran into a token limit during a long exchange. I noticed line 605 had a TODO to handle the token limit exceeded error. I figured one way to handle it was to drop the 3rd and 4th message from the messages array if the length was not less than 5. The idea being to keep the original request context, but to start dropping the earliest followup exchanges (eg "oh right do a pip install X")

## Related issue number

Did not open an issue.

## Checks

- N/A I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- N/A I've added tests (if relevant) corresponding to the changes introduced in this PR.
- N/A I've made sure all auto checks have passed.

## Other Notes

To get tests working I attempted to use that provided Docker image. But I had to do a ton of workaround to get it working out of the box. Had to do some [iffy fix to chroma](https://docs.trychroma.com/troubleshooting#sqlite) and had to install `pypdf` manually. So I am not sure if the tests would pass... but I am submitting this PR anyway for your review to at least consider my "just drop the second interaction" idea. 

Thanks for sharing `autogen` with the world!

### CODE REVIEW

**Comment by bioshazard** [comment]
@microsoft-github-policy-service agree

File: autogen/agentchat/conversable_agent.py
Comment (pcdeadeasy) [inline autogen/agentchat/conversable_agent.py:609]: Remove #Debug?

**Review by pcdeadeasy — CHANGES_REQUESTED** [review]
Looks code remove the extra code comments and print messages.

File: autogen/agentchat/conversable_agent.py
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:621]: Shall we raise InvalidRequestError?

Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:621]: Yes, that's OK.

Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py]: Delivered in e26f971

File: autogen/agentchat/conversable_agent.py
Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py]: delivered in e26f971

**Comment by bioshazard** [comment]
I updated the branch per the feedback. In my testing if you set the user proxy to NEVER ask the human, it is possible to get it snagged in a permanent cached loop. IMO that is expected behavior, but I figured to relay the experience. Thanks!

File: autogen/agentchat/conversable_agent.py
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:612]: The `InvalidRequestError` doesn't always mean context window overflow. Other type of `InvalidRequestError` shouldn't be handled by this approach.
Also, we should use this method only when users explicitly ask to. I think we can replace line 615 with a more generic function `compress_messages` which is configurable by users.
@kevin666aa what do you think?

Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py:612]: I probably should have looked at other PRs first. Is this made redundant by https://github.com/microsoft/autogen/pull/131 ? Or maybe still a valid stop-gap?

Comment (kevin666aa) [reply autogen/agentchat/conversable_agent.py:612]: In #131 , here is the usage of compression:
Before calling openai, a `on_oai_token_limit` is registered
- If compress_config set to false, the function would check if the existing token exceeds max token limit of a model. If so, it will terminate the conversation early to avoid token_limit_error.
- If compress_config set to true, the function would compress previous messages if user's preset token threshold is reached.

Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py:612]: It is not clear to me if any action should be taken on my end here. I intend to defer to the preference of the repo maintainers. Please let me know if there is anything specific I can contribute here. Thanks for the engagement so far y'all.

Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:612]: @kevin666aa is your PR compatible with this PR? How soon will you finish it? We have two paths now: merge this PR first (after necessary modification) and then base your PR off it; or merge your PR first and then base this PR of it. I generally prefer smaller, faster PRs than a giant and slow PR. Which PR requires lower effort to be ready to merge?

Comment (kevin666aa) [reply autogen/agentchat/conversable_agent.py:612]: This PR is compatible with mine. My PR handles messages between the request, and this PR handles it after the error. I think we can merge this if it is quicker.

Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:612]: Thanks. Could you suggest minimal changes required to merge this PR?

Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:612]: @kevin666aa

File: autogen/agentchat/conversable_agent.py
Comment (pcdeadeasy) [inline autogen/agentchat/conversable_agent.py:615]: The changes look fine but will be good to have a small comment to make the intent clear. This is done to overcome the token limit in case where len(messages) > 5 yet there is a **InvalidRequestError**

**Review by pcdeadeasy — COMMENTED** [review]
Please add comment to address the reason to remove message 2 and 3.

File: autogen/agentchat/conversable_agent.py
Comment (kevin666aa) [inline autogen/agentchat/conversable_agent.py:615]: This will delete pos 3 and 5 since the list will shift.
```suggestion
                del messages[2:4]
```

File: autogen/agentchat/conversable_agent.py
Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:612]: @kevin666aa what's your opinion about my comment:

The InvalidRequestError doesn't always mean context window overflow. Other type of InvalidRequestError shouldn't be handled by this approach.
Also, we should use this method only when users explicitly ask to. I think we can replace line 615 with a more generic function compress_messages which is configurable by users.

Comment (kevin666aa) [reply autogen/agentchat/conversable_agent.py:612]: I agree that `InvalidRequestError` is not equal to context window overflow, but I haven't see other types of  InvalidRequestError so far. This can be added as an option in Compression agent. I have changed it so that use can pass in different mode, current there are "TERMINATE" which terminates the chat, and "COMPRESS" . We can add a "DELETE" in it.

Comment (kevin666aa) [reply autogen/agentchat/conversable_agent.py:612]: > @kevin666aa what's your opinion about my comment:
> 
> The InvalidRequestError doesn't always mean context window overflow. Other type of InvalidRequestError shouldn't be handled by this approach. Also, we should use this method only when users explicitly ask to. I think we can replace line 615 with a more generic function compress_messages which is configurable by users.

I agree that `InvalidRequestError` is not equal to context window overflow, but I haven't see other types of  InvalidRequestError so far. This can be added as an option in Compression agent. I have changed it so that use can pass in different mode, current there are "TERMINATE" which terminates the chat, and "COMPRESS" . We can add a "DELETE" in it.

File: autogen/agentchat/conversable_agent.py
Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py]: Oh wow, you are right! Applied.

Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py]: Agreed. Added comments about why and how.

File: autogen/agentchat/conversable_agent.py
Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:612]: I have seen other types of InvalidRequestError. For example, when input arguments violate requirements.

Comment (bioshazard) [reply autogen/agentchat/conversable_agent.py:612]: Is there a way to differentiate based on the exception? Can we assume the message dialog won't exceed five by the time this error occurs? Functionally it does not seem to matter as my loop will terminate after enough messages are removed.

Comment (kevin666aa) [reply autogen/agentchat/conversable_agent.py:612]: Then it is better to add it as an option in compression agent. @bioshazard Can you take a look at #131 if you have time? It would be great if you can give it a test.

Also, from what I have now, you can add this PR in it quickly by add a mode "DELETE" in `compress_config`, and then implement the corresponding method in function `on_oai_token_limit`

**Comment by gagb** [comment]
@kevin666aa -- just gathering state of old PRs. My understanding is that this PR is outdated but the feature it is trying to implement will instead be addressed/implemented in https://github.com/microsoft/autogen/issues/156. Is my understanding correct?

**Comment by kevin666aa** [comment]
Yes, I think this PR is outdated, but the constrained message idea could be implemented as a mode in CompressibleAgent #443. I am closing this PR. @bioshazard It would be great if you can take a look at #443. The idea here is great and it can be implemented as a "WINDOW" mode in CompressibleAgent. Please let me know if you would like to make changes there.

**Comment by bioshazard** [comment]
@kevin666aa  @sonichi  not sure why but I am getting alerts for 30 day duration jobs in Actions for this repo? https://github.com/microsoft/autogen/actions/runs/6476770447

**Pull request closed without merging**
